### PR TITLE
Work on performance improvements

### PR DESCRIPTION
**Pull request by mdedetrich**
Ongoing work for #100

### CODE REVIEW

**Comment by cb372** [comment]
Thanks! But why are you doing if-null checks in some places and `try { ... } catch NPE` in others? Could we standardise on the former please. I think throwing and catching an exception is much slower than a null check.

**Comment by mdedetrich** [comment]
> But why are you doing if-null checks in some places and try { ... } catch NPE in others? 

According to the documentation for that specific library it throws a `NullPointerException`. It may be a case of me being stupid, is throwing a `NullPointerException` the same thing as returning `null` in Java land?

**Comment by cb372** [comment]
Neither Guava nor Caffeine will throw an NPE if the value is not present.

* [Javadoc for Guava's getIfPresent method](https://google.github.io/guava/releases/18.0/api/docs/com/google/common/cache/Cache.html#getIfPresent(java.lang.Object))
* [And for Caffeine's getIfPresent](http://static.javadoc.io/com.github.ben-manes.caffeine/caffeine/2.3.0/com/github/benmanes/caffeine/cache/Cache.html#getIfPresent-java.lang.Object-)

If the value is not present, both libraries will return null. So I'd recommend the following code for Guava and Caffeine:

```scala
val entry = underlying.getIfPresent(key).asInstanceOf[Entry[V]]
if (entry == null || entry.isExpired) None
else Some(entry.value)
```

(Note that calling `asInstanceOf[...]` on `null` is perfectly fine, it won't throw an NPE.)

File: ehcache/src/main/scala/scalacache/ehcache/EhcacheCache.scala [hunk 1]
```diff
@@ -28,12 +28,17 @@ class EhcacheCache(underlying: Ehcache)
...
-    val result = for {
-      e <- Option(underlying.get(key))
-      v <- Option(e.getObjectValue.asInstanceOf[V])
-    } yield v
-    logCacheHitOrMiss(key, result)
-    Future.successful(result)
+
+    val valueObj = underlying.get(key)
+
+    if (valueObj != null) {
```
Comment (cb372) [inline ehcache/src/main/scala/scalacache/ehcache/EhcacheCache.scala]: EhCache is a bit tricky. There are multiple cases to handle here:

* `underlying.get(key)` returns `null`
* It returns a non-null element, but `elem.getObjectValue` is `null`

I'd suggest this:

```scala
val result = {
  val elem = underlying.get(key)
  if (elem == null) None
  else Option(elem.getObjectValue.asInstanceOf[V])
}
logCacheHitOrMiss(key, result)
Future.successful(result)
```

**Comment by mdedetrich** [comment]
Hmm, I swear that the docs said it would throw a NullPointerException, let me look.

For the Caffeine Cache, the documentation for `getIfPresent` states this

```java
  /**
   * Returns the value associated with {@code key} in this cache, or {@code null} if there is no
   * cached value for {@code key}.
   *
   * @param key key whose associated value is to be returned
   * @return the value to which the specified key is mapped, or {@code null} if this map contains no
   *         mapping for the key
   * @throws NullPointerException if the specified key is null
   */
  @CheckForNull
  V getIfPresent(@Nonnull Object key);
```

So according to the docs it is meant to throw null?

I will look at the other changes when I get home tonight

**Comment by cb372** [comment]
To summarise the Javadoc you posted above:

* If there is nothing in the cache with the given key, it will *return* `null`
* If you pass it `null` as the key that you want to look up, it will *throw* a `NullPointerException`

We need to handle the first case by doing a null-check on the returned value.

We don't really care about the second case: If you try to look something up with a `null` key, you deserve a NullPointerException! 😛

**Comment by mdedetrich** [comment]
Ah thanks I misread the docs 😜 Will do the requested adjustments today!

**Comment by mdedetrich** [comment]
Did the requested changes, anything else that needs to be done?

**Comment by cb372** [comment]
Thanks, this looks fine now. I ran the micro-benchmark against this branch and it looks like there is a slight speed boost thanks to the reduction in object allocations 👍

**Pull request merged**